pkg/wrangler/capi.go

```diff
@@ -165,27 +190,37 @@ func (d *DeferredCAPIRegistration) DeferFuncWithError(clients *Context, f func(w
 }
 
 func (d *DeferredCAPIRegistration) DeferRegistration(ctx context.Context, clients *Context, register func(ctx context.Context, clients *Context) error) error {
-	d.lock.Lock()
-	defer d.lock.Unlock()
 	d.wg.Add(1)
-	if d.CAPIInitialized {
+	if d.CAPIInitialized() {
+		logrus.Infof("[deferred-capi - DeferRegistration] Executing deferred registration function as CAPI is initilized")
+		defer func() {
+			logrus.Infof("[deferred-capi - DeferRegistration] deferred registration function has completed")
+		}()
+
+		d.mutex.Lock()
+		defer d.mutex.Unlock()
 		return clients.StartSharedFactoryWithTransaction(ctx, func(ctx context.Context) error {
 			if err := d.invokeRegistrationFuncs(ctx, clients, []func(ctx context.Context, clients *Context) error{register}); err != nil {
 				return err
 			}
 			return nil
 		})
 	}
+
+	d.mutex.Lock()
 	d.registrationFuncs = append(d.registrationFuncs, register)
+	d.mutex.Unlock()
```

Comment:
This does not feel correct.
Ok, the `CAPIInitialized()` locks internally.
If it returns `true` we go to the execution of the function. And lock again. We have a small window where other things may happen to the structure.
Similarly, for `false` we lock again to extend the list of deferred functions.
And a small window where other things may happen to the structure. Like getting it initialized.
At which point the registered function may be lost and never called.

I believe that the original lock/unlock covering/protecting the entire function was more correct.
No window for other goroutines to interfere.

Ditto for `DeferFunc`. (Maybe also `DeferFuncWithError`, cannot see it in all the changes and comments).

Reply:
The new locking look mostly ok. You already saw my comments where things were missing, and addressed them :+1: